nixosTests: Reraise exceptions in subtests so they stop the test

tfc opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

#72828

While porting the `printing` test i realized that if an `assert` triggers within a `subtest`, then this is logged but does not lead to test abort. This was bad because without successful assertions the rest of the test didn't make any sense.

I changed this passage of the test driver so that it gives extra-info in which subtest the problem occured and then leave it to the global exception handling to fail the whole test.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

tfc committed:
nixosTests: Reraise exception in subtests

flokli commented:
This basically depends on whether we want to use subtests to describe individual independent things / testcases on machines that don't necessarily depend on each other (A), or just use them as a way to group / give a heading to a bunch of statements (B).

The current behaviour (both in python and perl) is to add the an error message to `failed_tests`, but keep going with the next one if I'm not mistaken (cc @globin https://github.com/NixOS/nixpkgs/pull/74089), which assumes (A)

However, I think we mostly do B in many tests, even though we probably shouldn't. WDYT?

I don't think we should change semantics here with the perl driver still around.

tfc commented:
> This basically depends on whether we want to use subtests to describe individual independent things / testcases on machines that don't necessarily depend on each other (A), or just use them as a way to group / give a heading to a bunch of statements (B).

It generally makes sense to enable the user to decide which things should make tests fail and which should not.

> The current behaviour (both in python and perl) is to add the an error message to `failed_tests`, but keep going with the next one if I'm not mistaken (cc @globin #74089), which assumes (A)

True, but i observed this behavior to do no good in the many many tests that i have ported.

> However, I think we mostly do B in many tests, even though we probably shouldn't. WDYT?

The vast majority of tests that i have seen assumes behavior B. Behavior B is also a good default setting in my opinion.

We could longterm give users some wrapper that is similar to `subtest` but that is explicit in isolating exceptions from the code outside so users can kind of add diagnostic output. This however with the current situation does not make much sense at the time because what kind of test should succeed if something in it failed?

flokli commented:
I agree that a lot of tests currenly assume B.

I'm just hesitant introducing semantic changes with both python and perl driver around, so would really feel better if we pick this up when the perl driver isn't around anymore.

tfc commented:
> I'm just hesitant introducing semantic changes with both python and perl driver around, so would really feel better if we pick this up when the perl driver isn't around anymore.

I agree that it is not cool to have different semantics in different implementations that share the same names.

But we have a lot of things in subtests that are necessary prerequisites to their next steps, but won't blow up visibly, leading to tests that time out (e.g. something must be done, but that fails, and then a test loops on `wait_until_x` which then can never happen) and are hard to debug.

Regarding the minority of still existing perl tests that remain untouched and working until they are ported (and people will see that they get more fails instead of less then, which is always better than the other way around), i suggest putting this in.

worldofpeace commented:
Just want to mention if there's any decision on this please merge after branch off.

flokli commented:
Sure.

flokli commented:
I stumbled more than once into python tests failing in a previous subtests, while I was staring at the later (still running) output.

As this only affects the behaviour of tests ultimately failing anyway, and tests run isolated too, I don't have any strong objections against this.

flokli committed:
nixosTests: drop nr_tests and failed_tests variables

With the tests now bailing out early on a failing subtest, we don't need
to keep a list of failed tests, or the number of total tests

flokli commented:
I pushed some cleanups, PTAL.

tfc commented:
@flokli looks good!
